ENH: Add support for inplace matrix multiplication

BvB93 opened the pull request:
This PR adds support for inplace matrix multiplication via the `@=` operator.

From what I gather, `@=` was made illegal back in the day when `np.matmul` was still  a normal function, as the latter lacked support for something akin to an `out` parameter and `a @= b` would thus be silently expanded to `a = a @ b`. This limitation was eventually resolved when matmul was converted into a ufunc, removing the issue previously plaguing `@=`.

Examples
----------
``` python
>>> import numpy as np
>>> a = np.arange(6).reshape(3, 2)
>>> print(a)
[[0 1]
 [2 3]
 [4 5]]

>>> b = np.ones((2, 2), dtype=int)
>>> a @= b
>>> print(a)
[[1 1]
 [5 5]
 [9 9]]
```

Commit:
ENH: Add support for inplace matrix multiplication

Commit:
DOC: Remove an outdated `matmul` comment

Commit:
TST: Add tests for inplace matrix multiplication

Commit:
MAINT: Let `ndarray.__imatmul__` handle inplace matrix multiplication in the array-api

seberg commented:
Could you add a test for a large matrix, I am slightly worried about this only accidentally working.  The problem is that if the input is exactly identical to the output, we probably pass this on into lapack (einsum?) directly.
Even after a quick check, I am not sure that the `matmul` inner-loop is actually able to deal with overlap just fine.

Commit:
TST: Add a dedicated `__imatmul__` test case for large matrices

BvB93 commented:
> Could you add a test for a large matrix, I am slightly worried about this only accidentally working. 

I just added a dedicated test case in https://github.com/numpy/numpy/pull/21120/commits/1e62199e5d5c5198684ad0ed359ad53eeb62050d; even when upping the size of the matrix from `10**6` to `10**8` I'm still getting the expected result. I assume these results are somewhat promising?

seberg commented:
Thanks, I like to have the test in either case, I will have a closer look at the code.

I have to check if and when the iterator makes a copy anyway.  If it does (which is plausible) writing `a @= b` vs. `a = a @ b` has no advantage (in fact it is probably worse), but it will be safe.  If it does not copy, things get a bit tricky, because `matmul` has a bunch of different code paths that may be taken, and we have to check/test all of them.

seberg commented:
Sorry, my bad, I missed/forgot that we of course already have a `matmul` special case that always forces the copy.  So this is safe, but slow/not very useful.

eric-wieser reviewed (commented):
Can you add a test for what happens when `a @ b` and `a` are not the same shape? Especially cases where a broadcast could happen, if `a` is `(n, n)` and `b` is `(n, 1)`.

seberg commented:
> Especially cases where a broadcast could happen, if a is (n, n) and b is (n, 1).

Great point, and potentially tricky :/.  I remember fixing some things around this and making it more strict, but we did *not* disallow the output to broadcast the inputs at that time.  This means that this is still incorrect:
```
In [4]: arr = np.array([1., 2.])

In [5]: np.matmul(arr, arr, out=arr)
Out[5]: array([5., 5.])
```
There could be an argument to allow the output to broadcast in some cases, but when inplace ops are used for a gufunc we are on _very_ thin ice.

BvB93 committed:
TST: Add `__imatmul__` tests for arrays with various shapes

BvB93 commented:
> Can you add a test for what happens when `a @ b` and `a` are not the same shape? Especially cases where a broadcast could happen, if `a` is `(n, n)` and `b` is `(n, 1)`.

Currently it raises an exception, rather than automatically broadcasting the output. In this sense it's somewhat unique compared to the other inplace operation, but then again `__matmul__` is also the only dunder that can produce an array that is smaller in size compared to the original.

All in all I think I might prefer the current (more explicit) `ValueError` over an implicit broadcast (this sounds like something that could easily bite someone in the ass).

``` python
In [1]: a = np.arange(9).reshape(3, 3)
   ...: b = np.ones((3, 1), dtype=int)

In [2]: a @= b
---------------------------------------------------------------------------
ValueError: matmul: Output operand 0 has a mismatch in its core dimension 1, with gufunc signature (n?,k),(k,m?)->(n?,m?) (size 3 is different from 1)
```

BvB93 commented:
On a unrelated note: either sphinx or one its dependencies is causing issues in circleci. 
Based on the exception a package requiring python >= 3.9 is being installed on python 3.8.

``` python
Traceback (most recent call last):
  File "/home/circleci/repo/venv/lib/python3.8/site-packages/sphinx/cmd/build.py", line 280, in build_main
    app.build(args.force_all, filenames)
  File "/home/circleci/repo/venv/lib/python3.8/site-packages/sphinx/application.py", line 343, in build
    self.builder.build_update()
  File "/home/circleci/repo/venv/lib/python3.8/site-packages/sphinx/builders/__init__.py", line 293, in build_update
    self.build(to_build,
  File "/home/circleci/repo/venv/lib/python3.8/site-packages/sphinx/builders/__init__.py", line 307, in build
    updated_docnames = set(self.read())
  File "/home/circleci/repo/venv/lib/python3.8/site-packages/sphinx/builders/__init__.py", line 412, in read
    self._read_parallel(docnames, nproc=self.app.parallel)
  File "/home/circleci/repo/venv/lib/python3.8/site-packages/sphinx/builders/__init__.py", line 463, in _read_parallel
    tasks.join()
  File "/home/circleci/repo/venv/lib/python3.8/site-packages/sphinx/util/parallel.py", line 108, in join
    if not self._join_one():
  File "/home/circleci/repo/venv/lib/python3.8/site-packages/sphinx/util/parallel.py", line 129, in _join_one
    raise SphinxParallelError(*result)
sphinx.errors.SphinxParallelError: AttributeError: 'PosixPath' object has no attribute 'readlink'
```

BvB93 commented:
>  I am not quite sure what to do about my example where the number of core dimensions of input and output differs; can we ignore it?

I would be in favor of ignoring it and just letting it raise (_i.e._ the current behavior). There a few other functions that reduce either the dimensionality or alter the output shape w.r.t. the input shape (_e.g._ `np.mean`) and those raise a `ValueError` as well, even though the result could in principle be broadcasted.

``` python
In [1]: a = np.arange(9).reshape(3, 3)
   ...: b = np.ones((3, 1), dtype=int)

In [2]: b.mean(out=a)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
  ...
ValueError: output parameter for reduction operation add has the wrong number of dimensions: Found 2 but expected 0
```

seberg commented:
@BvB93 sorry, I should have repeated my example:
```python
In [1]: arr = np.array([1.0, 2.0])

In [2]: arr @ arr
Out[2]: 5.0

In [3]: arr @= arr

In [4]: arr
Out[4]: array([5., 5.])
```

BvB93 committed:
MAINT: Explicitly raise when `a @= b` would otherwise broadcast the output

Add special casing for `1d @ 1d` and `2d @ 1d` ops.

seberg commented on numpy/core/src/multiarray/number.c:
```diff
@@ -348,13 +349,32 @@ array_matrix_multiply(PyObject *m1, PyObject *m2)
+    }
+
+    INPLACE_GIVE_UP_IF_NEEDED(m1, m2_array,
+            nb_inplace_matrix_multiply, array_inplace_matrix_multiply);
```
Yes, it is always done first so lets do it first here also.

seberg committed:
MAINT: Move check and expand error message slightly

In principle, this is probably still not 100% correct always since
we convert the the other object to an array upfront (to get the
error).  But the alternative solution using `axes=` seems tricky
as well...

seberg committed:
ENH: Implement matmul using the nuclear options

This uses the `axes` argument.  Arguably the most correct version since
we use the full ufunc machinery, so we can't mess up with random conversions
(which the test suite actually did notice, at least for an array subclass).

OTOH, for now the errors are ridiculously unhelpful, some of which could
be fixed in the gufunc code (at least made _better_).

seberg commented:
I pulled the nuclear option on this, because the tests noticed the array coercion.  This means that the errors are pretty terrible, maybe most impressive:
```
In [2]: arr = np.ones((3, 3))
In [3]: arr @= np.ones(3)
ValueError: axes item 1 should be a tuple with a single element, or an integer
```
but at least the `ValueError` part is technically correct :cough:.

mattip commented:
By "nuclear" you mean call back out to python with `np.matmul(a, b, out=a, axes=[(-2, -1), (-2, -1), (-2, -1)])`. How do you think we could get a better error message?

seberg commented:
Yeah, that is what I mean, the initial code converted the `b` to an array in `a @= b`, which makes it easy to check things up-front.

We can improve the generic message a little, but that will still be a bit confusing probably.  I suspect getting a nice one would actually require to manually replace it here (maybe pragmatically assuming that the other side isn't a super weird object).

seberg commented:
Something like "operand 2 has 1 dimension, but 2 dimensions are required (axes tuple has length 2)."

seberg committed:
ENH: Replace AxisError with more spefic one

In theory, an object matmul could be caught here, but lets assume
that doesn't happen...

seberg commented:
OK, now based on changing the other error to use AxisError, replacing it here now (a bit of a best effort since you could do evil things with nested object arrays).  That removes the worst of the errors (which referred to the `axes` argument the user never even supplied).

seberg committed:
BUG: Remove self == res assert for inplace matmul

It appears that assertion is not true for bad __array_ufunc__
or similar implementations.  And we have at least one test that runs
into it.

mattip commented:
I think we should put this in and work out better error messages as they arise. For the case above, the message is now:
```
>>> arr = np.ones((3, 3))
>>> arr @= np.ones(3)
Traceback (most recent call last):
  File "<console>", line 1, in <module>
ValueError: inplace matrix multiplication requires the first operand to have \
    at least one and the second at least two dimensions.
```

A casting error also seems clear enough:
```
>>> arr = np.ones((3, 3), dtype=int)
>>> arr @= np.ones((3,3))
Traceback (most recent call last):
  File "<console>", line 1, in <module>
numpy.core._exceptions._UFuncOutputCastingError: Cannot cast ufunc 'matmul' output \
    from dtype('float64') to dtype('int64') with casting rule 'same_kind'
```

seberg commented:
I finished this from my side, if you are happy with the new solution.

mattip commented:
Thanks @BvB93 and @seberg.
